doc/compiler/optimizer/ValuePropagation.md

```diff
@@ -0,0 +1,195 @@
+
+Because value propagation relies on all nodes having valid value numbers it is impossible for us to perform a transformation which creates a new node out of thin air during the analysis. This is because for loop back edges for example we may visit the same node twice, and we could potentially encounter a node without a value number. We could of course create a unique value number for this new node at the point of creation, but for one reason or another we chose to handle such operations by employing the notion of _delayed transformations_.
+
+During value propagation if we encounter a situation where we would like to create a brand new node, what we do is we record this node into a list for later processing. Once the the value propagation analysis is done we no longer have a need for the value numbers, and we can safely process all the recorded nodes. This operation is done by the [doDelayedTransformations](https://github.com/eclipse/omr/blob/25f479d2b671c2a0a03156cc0a48024a46d2e3c2/compiler/optimizer/OMRValuePropagation.hpp#L467) API.
+
+Note that if we want to transmute node `A` into node `B` we can still do so during the analysis phase as long as both node `A` and node `B` have valid value numbers. Similarly we can eliminate nodes from the program by simply removing them (for example eliminating `null` checks).
+
+## Handling Exception Edges
+
+In OMR exception edges do not break up basic blocks, instead they are handled via what are known as _check nodes_. This was a decision made very early in the development of the compiler component so as to avoid having control flow graphs with many tiny blocks which would hinder the effectiveness of local optimizations.
+
+These check nodes can be quite common in languages such as Java where we would have nodes representing null checks (`NULLCHK`), bound checks (`BNDCHK`), etc. All of these check nodes can cause an exception to occur (`NullPointerException` and `ArrayIndexOutOfBoundsException` respectively). In the compiler we create exception edges from blocks containing such excepting nodes to their corresponding catch blocks.
+
+Optimizations such as value propagation which walk the control flow graph need to deal with such exception edges. The reason for this is because execution can resume if the catch block handles the exception and falls through to the next block. We need to properly keep track of the constraints at the merge point of the catch block and the fall-through of the try block.
+
+To do this the value propagation optimization merges the set of constraints known at the time we are examining the check node onto the exception edge. Each new check node we encounter we perform another merge operation because the set of constraints may have changed between two check nodes which could be several tree tops apart in the _same_ block.
```

Comment:
This is accurate but I did want to state that there can be nodes other than check nodes that can cause exceptions. One case in point is call nodes. So this description may be better talking about nodes that "may cause an exception" rather than "check" nodes.

Reply:
Fixed in [force-push](https://github.com/eclipse/omr/compare/5c1aa757ca3e116307e60404b9d3b0743e2ce03f..127bdeabc6860a9f211446b05d6210a084512605).